hare3/hare.go

```diff
@@ -354,56 +377,67 @@ func (h *Hare) onLayer(layer types.LayerID) {
 	})
 }
 
-func (h *Hare) run(layer types.LayerID, beacon types.Beacon, proto *protocol) error {
+func (h *Hare) run(session *session) error {
 	// oracle may load non-negligible amount of data from disk
 	// we do it before preround starts, so that load can have some slack time
 	// before it needs to be used in validation
-	current := IterRound{Round: preround}
-
-	start := time.Now()
-	vrf := h.oracle.active(h.signer.NodeID(), layer, current)
+	var (
+		current = IterRound{Round: preround}
+		start   = time.Now()
+		active  bool
+	)
+	for i := range session.signers {
+		session.vrfs[i] = h.oracle.active(session.signers[i], session.beacon, session.lid, current)
+		active = active || session.vrfs[i] != nil
+	}
+	h.tracer.OnActive(session.vrfs)
 	activeLatency.Observe(time.Since(start).Seconds())
-	h.tracer.OnActive(vrf)
 
-	walltime := h.nodeclock.LayerToTime(layer).Add(h.config.PreroundDelay)
-	if vrf != nil {
-		h.log.Debug("active in preround", zap.Uint32("lid", layer.Uint32()))
+	walltime := h.nodeclock.LayerToTime(session.lid).Add(h.config.PreroundDelay)
+	if active {
+		h.log.Debug("active in preround", zap.Uint32("lid", session.lid.Uint32()))
 		// initial set is not needed if node is not active in preround
 		select {
 		case <-h.wallclock.After(walltime.Sub(h.wallclock.Now())):
 		case <-h.ctx.Done():
 			return h.ctx.Err()
 		}
 		start := time.Now()
-		proto.OnInitial(h.proposals(layer, beacon))
+		session.proto.OnInitial(h.proposals(session))
 		proposalsLatency.Observe(time.Since(start).Seconds())
 	}
-	if err := h.onOutput(layer, current, proto.Next(vrf != nil), vrf); err != nil {
+	if err := h.onOutput(session, current, session.proto.Next(active)); err != nil {
 		return err
 	}
 	result := false
 	for {
 		walltime = walltime.Add(h.config.RoundDuration)
-		current := proto.IterRound
-		var vrf *types.HareEligibility
-		if current.IsMessageRound() {
-			start := time.Now()
-			vrf = h.oracle.active(h.signer.NodeID(), layer, current)
-			activeLatency.Observe(time.Since(start).Seconds())
+		current := session.proto.IterRound
+
+		start := time.Now()
+		for i := range session.signers {
+			if current.IsMessageRound() {
+				session.vrfs[i] = h.oracle.active(session.signers[i], session.beacon, session.lid, current)
+			} else {
+				session.vrfs[i] = nil
+			}
+			active = active || session.vrfs[i] != nil
```

Comment:
Previously `active` was expressed as `vrf != nil`. And previously `vrf` was set fresh every for loop iteration. For example, it could theoretically be `!=nil` in iter 0 and `nil` in the next one. Now, the `active` remains `true` once it's set. Is it on purpose?

Reply:
i realized that i can simplify code a bit and drop `active` completely. code always generate a message and decide later whether to publish it or not. but i will reset active to false for now, so that it is consistent with previous behavior